Frameworks/Foundation/NSDecimal.mm

```diff
@@ -0,0 +1,605 @@
+    for (int i = length - 1; i >= 0; --i) {
+        result = i + 1;
+        if (mantissa[i] != 0) {
+            break;
+        }
+    }
+
+    return result;
+}
+
+/**
+ @Status Stub
+ @Notes
+*/
+NSCalculationError NSDecimalDivide(NSDecimal* result,
+                                   const NSDecimal* leftOperand,
+                                   const NSDecimal* rightOperand,
+                                   NSRoundingMode roundingMode) {
+    UNIMPLEMENTED();
+    return StubReturn();
+}
+
+static NSCalculationError _mantissaDivision(NSDecimalInternal* result, NSDecimalInternal* dividend, unsigned short divisor) {
+    // Note: All inputs have been normalized via NSDecimalNormalize
+    if (!divisor) {
+        INVALIDATE_DECIMAL((&result->decimal));
+        return NSCalculationDivideByZero;
+    }
+
+    for (int index = dividend->decimal._length - 1; index >= 0; --index) {
+        unsigned int res = static_cast<unsigned int>((dividend->_mantissa)[index]) | (result->_remainder << 16);
+        result->_remainder = res % divisor;
+        (result->_mantissa)[index] = res / divisor;
+    }
+
+    result->decimal._length = _obtainLength(result->_mantissa, dividend->decimal._length);
+
+    return NSCalculationNoError;
+}
+
+static NSCalculationError _compactMantissa(NSDecimalInternal* iDecimal, NSRoundingMode roundingMode) {
+    // Note: All inputs have been normalized via NSDecimalNormalize
+    // divide by 10 until we can compact it.
+    int exponent = 0;
+
+    NSDecimalInternal result;
+    while (iDecimal->decimal._length > NSDecimalMaxSize) {
+        exponent++;
+        NSCalculationError error = _mantissaDivision(iDecimal, iDecimal, 10);
+        if (error != NSCalculationNoError) {
+            return error;
+        }
+    }
+
+    // update the exp
+    result.decimal._exponent += exponent;
+
+    // TODO: Add Support for rounding
+    if (result._remainder != 0) {
+    }
+
+    return NSCalculationNoError;
+}
+
+/**
+ @Status Stub
+ @Notes
+*/
+NSString* NSDecimalString(const NSDecimal* dcm, id locale) STUB_METHOD {
+    UNIMPLEMENTED();
+    return StubReturn();
+}
+
+/**
+ @Status Interoperable
+*/
+NSComparisonResult NSDecimalCompare(const NSDecimal* leftOperand, const NSDecimal* rightOperand) {
+    if (leftOperand == rightOperand) {
+        return NSOrderedSame;
+    }
+
+    THROW_NS_IF_FALSE(E_INVALIDARG, (leftOperand != NULL) && (rightOperand != NULL));
+
+    NSDecimal left, right;
+    NSDecimalCopy(&left, leftOperand);
+    NSDecimalCopy(&right, rightOperand);
+    NSDecimalNormalize(&left, &right, 0);
+
+    if (!left._isNegative && right._isNegative) {
+        return NSOrderedDescending;
+    }
+
+    if (left._isNegative && !right._isNegative) {
+        return NSOrderedAscending;
+    }
+
+    // same signs
+    if (left._length < right._length) {
+        return right._isNegative ? NSOrderedDescending : NSOrderedAscending;
+    }
+
+    if (left._length > right._length) {
```

Comment:
>_length [](start = 29, length = 7)

What does `_length` represent? It would be _very_ helpful to have comments about the structure of NSDecimal and what you were intending for your algorithms.

If `_length` is a measure of the number of digits in the mantissa (or the number of `short`s required to store it), it is not correct to compare them to measure equality.

65538e-7 and 1e1 are Ascending, but you would say they are Descending simply because 65538 requires more `short`s to store than 1.